core/src/main/java/org/apache/metamodel/query/parser/FromItemParser.java

```diff
@@ -153,17 +183,23 @@ private SelectItem findSelectItem(String token, FromItem leftSide, FromItem righ
         if (result == null) {
             // fail over and try with the from items available in the join that
             // is being built.
-            final Query temporaryQuery = new Query().from(leftSide, rightSide);
```

Comment:
What I meant was that here you could just pass in the whole set of from items, instead of the two here. Then it's not more combinations, but fewer I think. At least you would then not need the new for-loop below.